### PR TITLE
feat(spv-790) Create outputs template and store evaluated template in destination

### PR DESCRIPTION
**Pull request by ac4ch**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/pike2/pike2.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	return []PikeOutputsTemplate{
+		{
+			Script:   scriptHex,
+			Satoshis: satoshis,
+		},
+	}, nil
+}
+
+// CalculatePubSharedSecret creates public shared secret base on sender public key
+func CalculatePubSharedSecret(senderPubKey *bec.PublicKey) *bec.PublicKey {
```
Comment (dorzepowski) [inline engine/pike2/pike2.go:66]: After some calculations it occurs, that this part (this calculation) makes no sense. Because in output we actually receive exactly the same public key. Sorry it was my misunderstanding of the concept.

Therefore calculating of the shared secret should be remove. And we should just use senderPublicKey instead in the later steps.

File: engine/pike2/pike2.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+	// Return the compressed and hashed public key
+	return dPKBytesCompressed
+}
+
+// GenerateLockingScriptsFromTemplates converts Pike outputs templates to scripts
+func GenerateLockingScriptsFromTemplates(outputsTemplate []PikeOutputsTemplate, senderPubKey, receiverPubKey *bec.PublicKey, reference string) ([]string, error) {
+	lockingScripts := make([]string, len(outputsTemplate))
+	parser := interpreter.DefaultOpcodeParser{}
+	pubSharedSecret := CalculatePubSharedSecret(senderPubKey)
```
Comment (dorzepowski) [inline engine/pike2/pike2.go:148]: As above. This step makes no sense as the output is exactly the same as senderPubKey.

File: engine/pike2/pike2.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		evaluated := make([]interpreter.ParsedOpcode, 0, len(parsedScript))
+		for _, op := range parsedScript {
+			switch op.Value() {
+			case bscript.OpPUBKEYHASH:
+				evaluated = append(evaluated, pkhParsed...)
+			case bscript.OpPUBKEY:
+				evaluated = append(evaluated, pkhParsed...) // Currently just appending the same key
+			default:
+				evaluated = append(evaluated, op)
+			}
```
Comment (dorzepowski) [inline engine/pike2/pike2.go:194]: It would be nice to have the whole logic of replacing templating OP codes in a separate function in package `script/template`.
The function could look like that:
```golang
func Evaluate(script []byte, pubKey *bec.PublicKey) []byte
```
or instead it could accept and return `string` (I don't know which version would be more sufficient to use in this code.

Also it would be good to have unit tests for it.

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/pike2/pike2.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+		bscript.OpCHECKSIG,
+	}
+
+	// Convert opcodes to hexadecimal string
+	scriptHex := hex.EncodeToString(opcodes)
+
+	// Check if the conversion was successful
+	if scriptHex == "" {
+		return nil, errors.New("failed to create script hex")
+	}
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/pike2/pike2.go:54]: done

File: engine/pike2/pike2.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+//
+// The package includes:
+// - Functions to generate P2PKH addresses from public keys.
+// - Methods to construct P2PKH scriptPubKey and scriptSig scripts.
+// - Utilities to decode and inspect P2PKH scripts and addresses.
+// - Validation functions to ensure the integrity and correctness of P2PKH scripts.
+//
+// This package is intended for developers working with Bitcoin or other cryptocurrencies that support P2PKH transactions.
+// It abstracts the low-level details and provides a high-level interface for creating and handling P2PKH scripts.
+package pike2
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/pike2/pike2.go:15]:
```suggestion
package pike
```
done

File: engine/pike2/pike2.go (hunk 1 above)
Comment (ac4ch) [reply engine/pike2/pike2.go:66]: simply changed to senderPubKey as we were doing multiplying by 1 with private key

File: engine/pike2/pike2.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+// newPrivateKey create a new private key with a given integer value
+func newPrivateKey(value int64) *bec.PrivateKey {
+	const privKeyBytesLen = 32
+	privKeyBytes := make([]byte, privKeyBytesLen)
+	bigValue := big.NewInt(value)
+	copy(privKeyBytes[privKeyBytesLen-len(bigValue.Bytes()):], bigValue.Bytes())
+	privateKey, _ := bec.PrivKeyFromBytes(bec.S256(), privKeyBytes)
+	return privateKey
+}
```
[the remark this answers is not in the document]
Comment (ac4ch) [reply engine/pike2/pike2.go:81]: removed

File: engine/pike2/pike2_test.go [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+				assert.NotNil(t, got)
+				if got != nil {
+					assert.True(t, bec.S256().IsOnCurve(got.X, got.Y))
+				}
+			}
+		})
+	}
+}
+
+func TestGenerateLockingScriptsFromTemplates(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ac4ch) [reply engine/pike2/pike2_test.go]: ok please have look now at these methods and let me know if we need to add this and to which function now :)

File: engine/pike2/pike2_test.go [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+	tests := []struct {
+		name            string
+		outputsTemplate []PikeOutputsTemplate
+		senderPubKey    *bec.PublicKey
+		receiverPubKey  *bec.PublicKey
+		reference       string
+		wantErr         bool
+	}{
+		{
+			name: "valid input",
```
[the remark this answers is not in the document]
Comment (ac4ch) [reply engine/pike2/pike2_test.go]: if we need more of this cases can u get me the keys .. i was able to find only 2 and i am using them already everywhere

File: engine/examples/client/pike/main.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+	"encoding/hex"
+	"fmt"
+
+	"github.com/libsv/go-bk/bec"
+
+	"github.com/bitcoin-sv/spv-wallet/engine/pike"
+	"github.com/bitcoin-sv/spv-wallet/engine/script/template"
+)
+
+func main() {
```
Comment (dorzepowski) [inline engine/examples/client/pike/main.go:13]: maybe it's good idea to change this into [testable example](https://go.dev/blog/examples) and put in the pike package?
We could even use the `Output` feature of testable examples to check the result of the method

File: engine/examples/client/pike/main.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+	receiverPubKey, err := bec.ParsePubKey(receiverPublicKeyBytes, bec.S256())
+	if err != nil {
+		panic(err)
+	}
+
+	// example of usage pike2
+	outputsTemplate, _ := template.P2PKH(10000)
+	fmt.Println(outputsTemplate)
+
+	lockingScripts, err := pike.GenerateLockingScriptsFromTemplates(outputsTemplate, senderPubKey, receiverPubKey, "reference")
```
Comment (dorzepowski) [inline engine/examples/client/pike/main.go:40]: It would be cool to have similar example (as testable example) for `template` package
so first create P2PKH template, and then call template.Evaluate on it with some public key.

File: engine/pike/pike.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"fmt"
+
+	"github.com/libsv/go-bk/bec"
+	"github.com/libsv/go-bt/v2/bscript"
+
+	"github.com/bitcoin-sv/spv-wallet/engine/script/template"
+	"github.com/bitcoin-sv/spv-wallet/engine/types/type42"
+)
+
```
Comment (dorzepowski) [inline engine/pike/pike.go:26]: IMHO we still should have a function `pike.GenerateOutputsTemplate(satoshis)`
which would just wrap call to `template.P2PKH(satoshis)` into a slice

File: engine/pike/pike_test.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	receiverPubKey, err := bec.ParsePubKey(receiverPubKeyBytes, bec.S256())
+	assert.NoError(t, err)
+
+	outputsTemplate := []template.P2PKHTemplate{
+		{Script: "76a914fd88ac"}, // Example P2PKH script in hex
+		{Script: "76a914fe88ad"}, // Another example P2PKH script in hex
+	}
+
+	t.Run("Valid Case", func(t *testing.T) {
+		lockingScripts, err := GenerateLockingScriptsFromTemplates(outputsTemplate, senderPubKey, receiverPubKey, "test-reference")
```
Comment (dorzepowski) [inline engine/pike/pike_test.go:33]: IMHO we should validate if lockingScripts, contains what we expect. I mean exactly those script strings that we know should be there.

File: engine/script/template/p2pkh_test.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+	errorTests := []struct {
+		name     string
+		satoshis uint64
+	}{
+		{
+			name:     "negative satoshis",
+			satoshis: ^uint64(0), // Simulating a case that would cause an error, maximum uint64 value, bitwise NOT of 0 is -1
+		},
+	}
```
Comment (dorzepowski) [inline engine/script/template/p2pkh_test.go:49]: Just a matter of taste,
but can you move it closer to the place where it it used

File: engine/script/template/p2pkh_test.go [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+				},
+			},
+		},
+		{
+			name:     "zero satoshis",
+			satoshis: 0,
+			expected: []P2PKHTemplate{
+				{
+					Script:   "76a9fd88ac",
+					Satoshis: 0,
```
Comment (dorzepowski) [inline engine/script/template/p2pkh_test.go:35]: can we assume that actually 0 is also invalid value?
Because this doesn't make sense to have Pay To Public Key Hash of 0 satoshis.

File: engine/script/template/p2pkh_test.go [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+			script:    []byte{bscript.OpDUP, bscript.OpHASH160, bscript.OpPUBKEY, bscript.OpEQUALVERIFY, bscript.OpCHECKSIG},
+			publicKey: mockPublicKey,
+			expected:  append([]byte{bscript.OpDUP, bscript.OpHASH160, bscript.OpDATA20}, append(mockPubKeyHash, bscript.OpEQUALVERIFY, bscript.OpCHECKSIG)...),
+		},
+		{
+			name:      "valid script without OP_PUBKEYHASH or OP_PUBKEY",
+			script:    []byte{bscript.OpDUP, bscript.OpHASH160, bscript.OpEQUALVERIFY, bscript.OpCHECKSIG},
+			publicKey: mockPublicKey,
+			expected:  []byte{bscript.OpDUP, bscript.OpHASH160, bscript.OpEQUALVERIFY, bscript.OpCHECKSIG},
+		},
```
Comment (dorzepowski) [inline engine/script/template/p2pkh_test.go:104]: Can you add another testcase, with the script 
OP_PUSHDATA1 **fd** OP_ADD OP_PUSHDATA1 **fe** OP_EQUALVERIFY
where fd and fe are hex data that match PUBKEY and PUBKEYHASH opcodes hexes but shouldn't be replaced by Evalueate, because it is only data.

File: engine/types/type42/linking_key_test.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/libsv/go-bk/bec"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCalculateHMAC(t *testing.T) {
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:12]: The question is if we really want to test this function, as this is private function. And actually it is kinda internal helper for DeriveLinkedKey, I would rather suggest to delete this tests entirely and just focus on tests for DeriveLinkedKey.

File: engine/types/type42/linking_key_test.go [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+		idx             int
+		expected        []byte
+		wantErr         bool
+	}{
+		{
+			name:            "valid input",
+			pubSharedSecret: []byte("shared_secret"),
+			reference:       "reference",
+			idx:             1,
+			expected:        computeExpectedHMAC([]byte("shared_secret"), "reference", 1),
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:26]: If we really want to test `calculateHMAC`, then we should have those expected values as static string, otherwise we can introduce bug in computeExpectedHMAC and get false positives or false negatives from those tests.

File: engine/types/type42/linking_key_test.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+				assert.Equal(t, tt.expected, got)
+			}
+		})
+	}
+}
+
+func TestDeriveLinkedKey(t *testing.T) {
+	sourcePubKeyHex := "027c1404c3ecb034053e6dd90bc68f7933284559c7d0763367584195a8796d9b0e"
+	sourcePubKeyBytes, err := hex.DecodeString(sourcePubKeyHex)
+	assert.NoError(t, err)
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:79]: It should fail, even before any test if we have invalid pubkey

```suggestion
	require.NoError(t, err)
```

The same change applies to all assert.NoError below

File: engine/types/type42/linking_key_test.go [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+
+	linkPubKeyHex := "03a34e456deecb6e6e9237e63e5b7d045d1d2a456eb6be43de1ec4e9ac9a07b50d"
+	linkPubKeyBytes, err := hex.DecodeString(linkPubKeyHex)
+	assert.NoError(t, err)
+	linkPubKey, err := bec.ParsePubKey(linkPubKeyBytes, bec.S256())
+	assert.NoError(t, err)
+
+	validHMAC, _ := calculateHMAC(sourcePubKey.SerialiseCompressed(), "valid-invoice")
+	validDerivedKey, _ := calculateDedicatedPublicKey(validHMAC, linkPubKey)
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:90]: We shouldn't calculate those in tests. 
We should just use a hardcoded strings with expected representation of the derived public key (hex of SerialiseCompressed)

File: engine/types/type42/linking_key_test.go [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+		{
+			name:          "invalid HMAC result",
+			source:        *sourcePubKey,
+			linkPubKey:    *linkPubKey,
+			invoiceNumber: "", // Empty invoice number causing HMAC calculation to fail
+		},
+		{
+			name:          "nil receiver public key",
+			source:        *sourcePubKey,
+			linkPubKey:    bec.PublicKey{}, // Empty public key causing dedicated public key calculation to fail
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:123]: It would be good to also provide a tests for `nil` used as source and used as linkPubKey

Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:123]: I wonder 🤔 why you use empty public key and not just `nil` ?

File: engine/types/type42/linking_key_test.go [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+		{
+			name:           "valid case",
+			source:         *sourcePubKey,
+			linkPubKey:     *linkPubKey,
+			invoiceNumber:  "valid-invoice",
+			expectedResult: validDerivedKey,
+		},
+	}
+
+	errorTests := []struct {
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:108]: It's a matter of taste,
but please move it closer to where it is used

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: engine/types/type42/linking_key.go [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+	// Check for nil receiver public key
+	if source.X == nil || source.Y == nil {
+		return nil, errors.New("source public key is nil")
+	}
+	if linkPubKey.X == nil || linkPubKey.Y == nil {
+		return nil, errors.New("receiver public key is nil")
+	}
+
+	// Compute the shared secret
+	sharedSecret := source.SerialiseCompressed()
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/types/type42/linking_key.go:82]:
```suggestion
	publicKeyBytes := source.SerialiseCompressed()
```
done

File: engine/types/type42/linking_key.go [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+
+	// Compute the HMAC result
+	hmacResult, err := calculateHMAC(sharedSecret, invoiceNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate the dedicated public key
+	dPK, err := calculateDedicatedPublicKey(hmacResult, &linkPubKey)
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/types/type42/linking_key.go:91]: ok
```suggestion
	linkedPK, err := calculateDedicatedPublicKey(hmacResult, &linkPubKey)
```

File: engine/types/type42/linking_key.go [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+	if _, err := h.Write([]byte(reference)); err != nil {
+		return nil, fmt.Errorf("error writing HMAC message - %w", err)
+	}
+	return h.Sum(nil), nil
+}
+
+// calculateDedicatedPublicKey calculates the dedicated public key (dPK) using the HMAC result and the receiver's public key.
+// The HMAC result is used as a scalar to perform elliptic curve scalar multiplication and point addition.
+// Returns the resulting public key or an error if the calculation fails.
+func calculateDedicatedPublicKey(hmacResult []byte, receiverPubKey *bec.PublicKey) (*bec.PublicKey, error) {
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/types/type42/linking_key.go:37]: ok

```suggestion
func calculateLinkedPublicKey(hmacResult []byte, receiverPubKey *bec.PublicKey) (*bec.PublicKey, error) {
```

File: engine/types/type42/linking_key.go [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+// The reference string, which acts as the HMAC key, can be an invoice number or any other string.
+// The function returns the HMAC result as a byte slice.
+//
+// Parameters:
+// - pubSharedSecret: A byte slice representing the public shared secret to be hashed.
+// - reference: A string used as the HMAC key.
+//
+// Returns:
+// - A byte slice containing the HMAC result as a byte slice or an error if the HMAC calculation fails.
+func calculateHMAC(pubSharedSecret []byte, reference string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/types/type42/linking_key.go:23]: ok
```suggestion
func calculateHMAC(pubSharedSecret []byte, message string) ([]byte, error) {
```

File: engine/pike/pike.go (hunk 11 above)
Comment (nonus25) [reply engine/pike/pike.go:26]: done
```suggestion
func GenerateOutputsTemplate(satoshis uint64) ([]template.P2PKHTemplate, error) {
	return template.P2PKH(satoshis)
}
```

File: engine/script/template/p2pkh.go [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+		if op.Value() == 0xFF {
+			return nil
+		}
+	}
+
+	// Serialize the public key to compressed format
+	dPKBytes := dPK.SerialiseCompressed()
+
+	// Apply Hash160 (SHA-256 followed by RIPEMD-160) to the compressed public key
+	dPKBytesCompressed := crypto.Hash160(dPKBytes)
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:82]: ok
```suggestion
	dPKHash := crypto.Hash160(dPKBytes)
```

File: engine/script/template/p2pkh.go [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+			evaluated = append(evaluated, pkhParsed...) // Currently just appending the same key
+		default:
+			evaluated = append(evaluated, op)
+		}
+	}
+
+	// Unparse the evaluated opcodes back into a script
+	finalScript, err := parser.Unparse(evaluated)
+	if err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:112]:
```suggestion
		return nil, err
```

File: engine/script/template/p2pkh.go [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+	}, nil
+}
+
+// Evaluate processes a given Bitcoin script by parsing it, replacing certain opcodes
+// with the public key hash, and returning the resulting script as a byte array.
+// Will replace any OP_PUBKEYHASH or OP_PUBKEY
+//
+// Parameters:
+// - script: A byte array representing the input script.
+// - dPK: A pointer to a bec.PublicKey which provides the dedicated public key to be used in the evaluation.
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:58]:
```suggestion
// - pubKey: A pointer to a bec.PublicKey which provides the dedicated public key to be used in the evaluation.
```
```suggestion
func Evaluate(script []byte, pubKey *bec.PublicKey) ([]byte, error) {
```

File: engine/script/template/p2pkh.go [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+	if scriptHex == "" {
+		return nil, errors.New("failed to create script hex")
+	}
+
+	// Create and return the PikeOutputsTemplate
+	return []P2PKHTemplate{
+		{
+			Script:   scriptHex,
+			Satoshis: satoshis,
+		},
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:48]: ok
```suggestion
	return &P2PKHTemplate{
		Script:   scriptHex,
		Satoshis: satoshis,
	}, nil
}
```

File: engine/script/template/p2pkh.go [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+// with the public key hash, and returning the resulting script as a byte array.
+// Will replace any OP_PUBKEYHASH or OP_PUBKEY
+//
+// Parameters:
+// - script: A byte array representing the input script.
+// - dPK: A pointer to a bec.PublicKey which provides the dedicated public key to be used in the evaluation.
+//
+// Returns:
+// - A byte array representing the evaluated script, or nil if an error occurs.
+func Evaluate(script []byte, dPK *bec.PublicKey) []byte {
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:62]: moved

File: engine/types/type42/linking_key_test.go (hunk 18 above)
Comment (nonus25) [reply engine/types/type42/linking_key_test.go:79]:
```suggestion
			assert "github.com/stretchr/testify/require"
```
done

File: engine/types/type42/linking_key_test.go (hunk 16 above)
Comment (nonus25) [reply engine/types/type42/linking_key_test.go:12]: easier to remove it, so remove it .. as anyway latter in code we are testing DeriveLinkedKey and if this private method would fail this method will fail to `DeriveLinkedKey`

File: engine/types/type42/linking_key_test.go (hunk 20 above)
Comment (nonus25) [reply engine/types/type42/linking_key_test.go:123]: because param function was not taking pointer but pubkey obj .. i will correct

File: engine/script/template/p2pkh.go [hunk 31]
```diff
@@ -0,0 +1,117 @@
...
+	"errors"
+
+	"github.com/libsv/go-bk/bec"
+	"github.com/libsv/go-bk/crypto"
+	"github.com/libsv/go-bt/v2/bscript"
+	"github.com/libsv/go-bt/v2/bscript/interpreter"
+)
+
+// P2PKHTemplate represents the script and satoshis for a Pike output
+type P2PKHTemplate struct {
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/script/template/p2pkh.go:15]:
```suggestion
type OutputTemplate struct {
	Script   string `json:"script"`
	Satoshis uint64 `json:"satoshis"`
}
```

File: engine/script/template/p2pkh_test.go (hunk 14 above)
Comment (nonus25) [reply engine/script/template/p2pkh_test.go:35]: ok

File: engine/examples/client/pike/main.go (hunk 9 above)
Comment (nonus25) [reply engine/examples/client/pike/main.go:13]: done

File: engine/examples/client/pike/main.go (hunk 10 above)
Comment (nonus25) [reply engine/examples/client/pike/main.go:40]: but we need 2 pub keys for this, anyway have look as i make this testable example

File: engine/examples/client/pike/main.go [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+	receiverPublicKeyBytes, err := hex.DecodeString(receiverPublicKeyHex)
+	if err != nil {
+		panic(err)
+	}
+	receiverPubKey, err := bec.ParsePubKey(receiverPublicKeyBytes, bec.S256())
+	if err != nil {
+		panic(err)
+	}
+
+	// example of usage pike2
```
[the remark this answers is not in the document]
Comment (nonus25) [reply engine/examples/client/pike/main.go:36]:
```suggestion
	// example of usage pike.GenerateOutputsTemplate
```

File: engine/script/template/evaluate.go [hunk 33]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	// Replace OP_PUBKEYHASH with the actual public key hash
+	evaluated := make([]interpreter.ParsedOpcode, 0, len(parsedScript))
+	for _, op := range parsedScript {
+		switch op.Value() {
+		case bscript.OpPUBKEYHASH:
+			evaluated = append(evaluated, pkhParsed...)
+		case bscript.OpPUBKEY:
+			return nil, errors.New("OP_PUBKEY not supported yet")
```
[the remark this answers is not in the document]
Comment (ac4ch) [reply engine/script/template/evaluate.go:63]: yes he does

File: engine/types/type42/linking_key_test.go [hunk 34]
```diff
@@ -0,0 +1,86 @@
...
+	linkPubKeyHex := "03a34e456deecb6e6e9237e63e5b7d045d1d2a456eb6be43de1ec4e9ac9a07b50d"
+	linkPubKeyBytes, err := hex.DecodeString(linkPubKeyHex)
+	assert.NoError(t, err)
+	linkPubKey, err := bec.ParsePubKey(linkPubKeyBytes, bec.S256())
+	assert.NoError(t, err)
+
+	validHMAC, err := calculateHMAC(sourcePubKey.SerialiseCompressed(), "valid-invoice")
+	assert.NoError(t, err)
+	validDerivedKey, err := calculateLinkedPublicKey(validHMAC, linkPubKey)
+	assert.NoError(t, err)
```
Comment (dorzepowski) [inline engine/types/type42/linking_key_test.go:27]: Still I don't like this approach to tests.
IMHO, we should have list of cases, where we have sender and recipient public keys, invoice number and expected derivedPublicKey.

And this is list of test cases that I generated from code from ts-sdk adjusted to our solution for derivations.
```
[
  {
    senderPublicKey: '033f9160df035156f1c48e75eae99914fa1a1546bec19781e8eddb900200bff9d1',
    recipientPublicKey: '02133b035cda4ba15f93b5fdde11c1f73eb9f1a79b60c6caa1c78e1c4c64ed72ce',
    invoiceNumber: 'f3WCaUmnN9U=',
    derivedPublicKey: '02d4049747553b4b956a419a7e1ddef418d57f317de8cc7d024d05c75f29309f26',
  },
  {
    senderPublicKey: '027775fa43959548497eb510541ac34b01d5ee9ea768de74244a4a25f7b60fae8d',
    recipientPublicKey: '02dfcbe35d95b55b5f3168ea8f12717e266ceddf88d04d2ff741272dfb0e542c2a',
    invoiceNumber: '2Ska++APzEc=',
    derivedPublicKey: '02679645bc44a771154f66eb52ca93507c3e777997a774e87611a7b17d30d748ee',
  },
  {
    senderPublicKey: '0338d2e0d12ba645578b0955026ee7554889ae4c530bd7a3b6f688233d763e169f',
    recipientPublicKey: '023c1db4cce57a44c95e05f0ed499085a9ce8ac83bf80dd4c7b4658a9d6c4a122e',
    invoiceNumber: 'cN/yQ7+k7pg=',
    derivedPublicKey: '03c07884a8e8f02bb1cfb276edddfa8dade7654d8cb744e0ff9e4b0c5dffa63ae0',
  },
  {
    senderPublicKey: '02830212a32a47e68b98d477000bde08cb916f4d44ef49d47ccd4918d9aaabe9c8',
    recipientPublicKey: '0363ec92e374974e9c904875ff8daff50bfa22d8ec0a39891a31e5e760aca5e9cf',
    invoiceNumber: 'm2/QAsmwaA4=',
    derivedPublicKey: '024d369f8440e91fdd58577bd6bc471d868fdfc58e605eb2ec5a2362df589d43cd',
  },
  {
    senderPublicKey: '03f20a7e71c4b276753969e8b7e8b67e2dbafc3958d66ecba98dedc60a6615336d',
    recipientPublicKey: '02cef40bd6826d1b1960bc94094c0c0a19547b291c33b7a07380cbdc4580f3678b',
    invoiceNumber: 'jgpUIjWFlVQ=',
    derivedPublicKey: '03aa3ea0de12f44642a51593b2f8a0e0ff9813cdb4ab39ab8039654dce2ae5bdb3',
  },
  {
    senderPublicKey: '03f20a7e71c4b276753969e8b7e8b67e2dbafc3958d66ecba98dedc60a6615336d',
    recipientPublicKey: '0363ec92e374974e9c904875ff8daff50bfa22d8ec0a39891a31e5e760aca5e9cf',
    invoiceNumber: 'jgpUIjWFlVQ=',
    derivedPublicKey: '032ac4dccb237d777361a7fade10fc2d21b642df9b0e4634d3e27fdce131526e4f',
  },
  {
    senderPublicKey: '02830212a32a47e68b98d477000bde08cb916f4d44ef49d47ccd4918d9aaabe9c8',
    recipientPublicKey: '02cef40bd6826d1b1960bc94094c0c0a19547b291c33b7a07380cbdc4580f3678b',
    invoiceNumber: 'jgpUIjWFlVQ=',
    derivedPublicKey: '02caa04e2020a0642e5283c978257d2135c00882c049adcd71a71dffe9d7208a39',
  },
  {
    senderPublicKey: '02830212a32a47e68b98d477000bde08cb916f4d44ef49d47ccd4918d9aaabe9c8',
    recipientPublicKey: '02cef40bd6826d1b1960bc94094c0c0a19547b291c33b7a07380cbdc4580f3678b',
    invoiceNumber: 'm2/QAsmwaA4=',
    derivedPublicKey: '030467472149ac58d9d04e4182b03af99593a7af312623c4be7d96f2fde08f6421',
  },
]
```
Public keys are serialised compressed. I hope I did it correctly.

**Review by dorzepowski — CHANGES_REQUESTED** [review]

**Review by chris-4chain — APPROVED** [review]

**Review by dorzepowski — APPROVED** [review]
